xlive/H2MOD/Modules/Shell/JsonConfig/JsonConfig.cpp

```diff
@@ -0,0 +1,253 @@
+                throw std::runtime_error("Failed to create file");
+            }
+            ofs << "{}";  // Initialize with an empty JSON object.
+            ofs.close();
+            ifs.open(filename_);
+        }
+        IStreamWrapper isw(ifs);
+        Document d;
+        doc_.SetObject();
+        if (doc_.ParseStream(isw).HasParseError())
+            return json_config_error::json_parse_error;
+
+        return json_config_error::success;
+    }
+
+    void save() {
+        std::ofstream ofs(filename_);
+        OStreamWrapper osw(ofs);
+        PrettyWriter<OStreamWrapper> writer(osw);
+        writer.SetIndent('\t', 1);  // Set indentation to use tabs.
+        doc_.Accept(writer);
+    }
+
+    Value* get_current_pointer()
+    {
+        // Check if path is empty, if so, return a pointer to the root document object
+        if (key_path.empty()) {
+            return Pointer("").Get(doc_);
+        }
+
+        // Obtain a pointer to the root document object
+        Value* value = Pointer("").Get(doc_);
+
+        // Iterate through each element in the path vector
+        for (const auto& element : key_path)
+        {
+            // Check if the current value is an object
+            if (value->IsObject()) {
+                // If the current element is not a member of the object, add a new member with the given name
+                if (!value->HasMember(element.c_str())) {
+                    Value key;
+                    key.SetString(element.c_str(), doc_.GetAllocator());
+                    value->AddMember(key, rapidjson::Value(rapidjson::kObjectType), doc_.GetAllocator());
+                }
+                // Update the value pointer to point to the member with the given name
+                value = &((*value)[element.c_str()]);
+            }
+            else {
+                // If the current value is not an object, return a null pointer
+                return nullptr;
+            }
+        }
+
+        // Return a pointer to the final value object in the path
+        return value;
+    }
+
+    json_config& operator[](const char* key)
+    {
+        key_path.push_back(key);
+        return *this;
+    }
+
+    // A Map to store default variables for the current instance of the class.
+    //
+    // json_config h2config(config.json);
+    // h2config.defaultValues_["crosshair_offset"] = NaN
+    //
+    std::unordered_map<std::string, Value> defaultValues_;
+
+
+
+    template<typename T>
+    T get(const char* key, T defaultValue = T{}) {
+        // Check if the key exists in the document
+
+        Value* current_object = get_current_pointer();
+        key_path.clear();
+
+        Value& v = (*current_object)[key];
+        if (!current_object->HasMember(key)) {
+            if (defaultValues_.find(key) != defaultValues_.end())
+                v = defaultValues_[key];
+            else
+                return defaultValue;
+        }
+
+    	if constexpr (std::is_same_v<T, bool>) {
```

Comment:
I think you can remove all of these `constexpr` specializations by simply using `return v.Get<T>()`. And some validation wouldn't hurt because just the type alone doesn't guarantee that the value/object in the json document is of the same type.

Reply:
Get<T> had compilation issues, I'll revisit it.

Reply:
Merged all of the base RapidJson value types into a single Get<T> call while leaving the expressions for the custom types.

Added Json Value validation that will check if there was a default parameter passed to it or return from the defaultValue store if it doesn't exist. If neither exist it will just return the default value of the specific type.